### PR TITLE
Avoid optimization of gcc (work around)

### LINKED ISSUE #519: Python problem on OSX
Hello,

Python examples are producing Segmentation fault error.  

~/LightGBM/examples/python-guide (master)$ python simple_example.py
Load data...
Start training...
Segmentation fault: 11

I can run examples using lightgbm image. 
LightGBM/examples/binary_classification (master)$ ../../lightgbm config=train.conf

## Environment info
OSX Siera
g++ (MacPorts gcc47 4.7.4_5) 4.7.4
Python 2.7.12 :: Anaconda 2.4.1 (x86_64

### LINKED ISSUE #521: GCC7 Problem: Check failed: offset == num_total_bin at feature_histogram.hpp
## Environment info
Operating System: macOS Sierra
CPU: i7 2.8Ghz
C++/Python/R version: Anaconda Python 3.5 (inside Jupyter Notebook) and GCC/G++ 7.1.0

## Error Message:

[LightGBM] [Warning] Check failed: offset == num_total_bin at /Users/.../LightGBM/src/treelearner/feature_histogram.hpp, line 484 .

terminate called after throwing an instance of 'std::runtime_error'
  what():  Check failed: offset == num_total_bin at /Users/.../LightGBM/src/treelearner/feature_histogram.hpp, line 484 .

## Code:

```
params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'num_leaves': 31,
    'learning_rate': 0.1,
    'feature_fraction': 0.7,
    'bagging_fraction': 0.7,
    'verbose': 0,
    'scale_pos_weight': 0.5,
    'min_sum_hessian_in_leaf': 1}


num_rounds=10
kf = KFold(n_splits=5, random_state=0, shuffle=True)
loss_l = []

for tr, ts in kf.split(X, y):
    xtr = lgb.Dataset(X[tr, :], label=y[tr], free_raw_data=False)
    xts = lgb.Dataset(X[ts, :], label=y[ts], free_raw_data=False, reference=xtr)

    gbm = lgb.train(params,
                    xtr,
                    num_boost_round=num_rounds)

    p = gbm.predict(xts)
    loss = log_loss(y[ts], p)
```

## Reproducible example

Running advanced_example.py gives me this:
Load data...
Start training...
python(4737,0x7fffb9b1a3c0) malloc: *** error for object 0x1001b76b0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6

Tried: 
- rebuilding lightgbm
- using data as numpy array, csr_matrix
- removing nans
- free_raw_data=False/True
- creating a copy of the slices of the original array (instead of view)

It creates the Dataset without any problems, but fails when I call lgb.train. 

**EDIT: Building with GCC 6.2.0 fixed the problem. But I still wanted to leave this open to report the problem with GCC7.**

Thanks!

### PR DESCRIPTION
**Pull request by henry0312**
Close #519
Close #521

### CODE REVIEW

**Comment by guolinke** [comment]
@henry0312 can you give some explanation ?

**Comment by henry0312** [comment]
I don't know why gcc-7.1.0's optimization breaks, but I found a bug when I changed dataset.cpp like this:

```diff
diff --git a/src/io/dataset.cpp b/src/io/dataset.cpp
index 80f059a..84f4a7a 100644
--- a/src/io/dataset.cpp
+++ b/src/io/dataset.cpp
@@ -197,13 +197,22 @@ std::vector<std::vector<int>> FastFeatureBundling(std::vector<std::unique_ptr<Bi
       }
     }
   }
+  for (size_t i = 0; i < ret.size(); ++i) {
+      for (size_t j = 0; j < ret[i].size(); ++j) {
+          std::cout << ret[i][j] << ' ';
+      }
+      std::cout << std::endl;
+  }
+  std::cout << "--------" << std::endl;
   // shuffle groups
   int num_group = static_cast<int>(ret.size());
   Random tmp_rand(12);
-  for (int i = 0; i < num_group - 1; ++i) {
-    int j = tmp_rand.NextShort(i + 1, num_group);
+  for (volatile int i = 0; i < num_group - 1; ++i) {
+    volatile const int j = tmp_rand.NextShort(i + 1, num_group);
+    std::cout << num_group << " " << i << " " << j << std::endl;
     std::swap(ret[i], ret[j]);
   }
+  std::cout << "--------" << std::endl;
   return ret;
 }
```

outputs:

```
❯ OMP_NUM_THREADS=1 python simple_example.py
Load data...
Start training...
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
--------
0 24
1 14
2 10
3 16
4 -17
5 -12
6 -11
7 6
8 8
9 25
10 0
11 1
12 5
13 9
14 27
15 26
16 18
17 20
18 19
19 20
20 16
21 22
22 20
23 25
24 25
25 26
26 27
--------
```

Obviously, -17, -12 and -11 are wrong.

**Comment by wxchan** [comment]
Isn't this just because `(RandInt16()) % (upper_bound - lower_bound)` can be negative?  https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/utils/random.h#L38

**Comment by henry0312** [comment]
No.
This is called at https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L204, and upper_bound should be greater than lower_bound.

**Comment by wxchan** [comment]
As long as `RandInt16()` can be negative, `(RandInt16()) % (upper_bound - lower_bound)` can be negative too.

http://stackoverflow.com/questions/7594508/modulo-operator-with-negative-values

**Comment by henry0312** [comment]
Ah, you are right.
After optimization of `x`, `RandInt16()` can be negative.

**Comment by guolinke** [comment]
@wxchan @henry0312 
here have a mask to make sure the return value is positive:
https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/utils/random.h#L98

**Comment by henry0312** [comment]
@guolinke hum...
Anyway, this PR gives a wrok-around.

**Comment by wxchan** [comment]
I try to compile with gcc-7, and I didn't get negative number for RandInt16(). Did you change other codes?

```
x:5099167 rand:77
0 24
x:368864998 rand:5628
1 14
x:408447505 rand:6232
2 10
x:1904010384 rand:29052
3 16
x:-2045365997 rand:1558
4 22
x:-433711222 rand:26150
5 20
x:-1298989019 rand:12947
6 18
x:-186224044 rand:29926
7 14
x:-1362257977 rand:11981
8 20
x:1466151790 rand:22371
9 25
x:-1880179591 rand:4078
10 26
x:-271217320 rand:28629
11 17
x:-1841736005 rand:4665
12 13
x:-2001385838 rand:2229
13 17
x:1618711309 rand:24699
14 27
x:1793743260 rand:27370
15 26
x:201916911 rand:3081
16 18
x:1180439798 rand:18012
17 20
x:-511330335 rand:24965
18 27
x:135281440 rand:2064
19 20
x:-385192605 rand:26890
20 24
x:1379836570 rand:21054
21 22
x:-1805017355 rand:5225
22 23
x:771872228 rand:11777
23 25
x:1956490519 rand:29853
24 25
x:-1953713282 rand:2956
25 26
x:-676856759 rand:22439
26 27
```

**Comment by henry0312** [comment]
@wxchan really?
Did you confirm #521 before and after this PR?
And did you try the change of https://github.com/Microsoft/LightGBM/pull/529#issuecomment-302348402?

**Comment by wxchan** [comment]
I change codes in dataset.cpp like this:
```
// shuffle groups
  int num_group = static_cast<int>(ret.size());
  Random tmp_rand(12);
  for (int i = 0; i < num_group - 1; ++i) {
    int rand_num = tmp_rand.RandInt16();
    int j = rand_num % (num_group - i - 1) + (i + 1);
    std::cout << "RandInt16:" << rand_num << " j:" << j << " x:" << tmp_rand.x << std::endl;
    std::swap(ret[i], ret[j]);
  }
  return ret;
```
The output is:
```
RandInt16:77 j:24 x:5099167
RandInt16:5628 j:14 x:368864998
RandInt16:6232 j:10 x:408447505
RandInt16:29052 j:16 x:1904010384
RandInt16:-31210 j:-17 x:-2045365997
RandInt16:-6618 j:-12 x:-433711222
RandInt16:-19821 j:-11 x:-1298989019
RandInt16:-2842 j:6 x:-186224044
RandInt16:-20787 j:8 x:-1362257977
RandInt16:22371 j:25 x:1466151790
RandInt16:-28690 j:0 x:-1880179591
RandInt16:-4139 j:1 x:-271217320
RandInt16:-28103 j:5 x:-1841736005
RandInt16:-30539 j:9 x:-2001385838
RandInt16:24699 j:27 x:1618711309
RandInt16:27370 j:26 x:1793743260
RandInt16:3081 j:18 x:201916911
RandInt16:18012 j:20 x:1180439798
RandInt16:-7803 j:19 x:-511330335
RandInt16:2064 j:20 x:135281440
RandInt16:-5878 j:16 x:-385192605
RandInt16:21054 j:22 x:1379836570
RandInt16:-27543 j:20 x:-1805017355
RandInt16:11777 j:25 x:771872228
RandInt16:29853 j:25 x:1956490519
RandInt16:-29812 j:26 x:-1953713282
RandInt16:-10329 j:27 x:-676856759
```
Look like `& 0x7fff` in RandInt16 somehow not be calculated in some cases.

**Comment by henry0312** [comment]
@wxchan Did you check that this PR prevented `RandInt16` from being negative?
